File: IdeaProjects/Memory/src/MemoryManager.java
import java.util.LinkedList;

final class FreeBlock{
    private int size;
    private boolean isAlloced;
    private int startAddr;

    FreeBlock(int size,boolean isAlloced,int startAddr){
        this.size = size;
        this.isAlloced = isAlloced;
        this.startAddr = startAddr;
    }

    FreeBlock(){
        this.size = 0;
        this.isAlloced = false;
        this.startAddr = 0;
    }


    public int getSize() {
        return size;
    }

    public void setSize(int size) {
        this.size = size;
    }

    public boolean isAlloced() {
        return isAlloced;
    }

    public void setAlloced(boolean alloced) {
        isAlloced = alloced;
    }

    public int getStartAddr() {
        return startAddr;
    }

    public void setStartAddr(int startAddr) {
        this.startAddr = startAddr;
    }
}


public abstract class MemoryManager {
    protected LinkedList<FreeBlock> freeBlockList;
    public abstract boolean MemoryAlloc(int size);
    public abstract void MemoryFree(FreeBlock freeBlock);
    public abstract void MemoryInit(int size);
    public abstract void showMemory();
}

File: CLionProjects/Memory/MemoryManager.cpp
//
// Created by zhuxingtian on 19-6-7.
//
#include "MemoryManager.h"

FreeBlock::FreeBlock(int size,bool Alloced,int startAddr)
        :size(size),Alloced(Alloced),startAddr(startAddr)
{
}

bool FreeBlock::operator==(const FreeBlock & other) const {
    return  (this->getSize() == other.getSize() &&
             this->getAlloced() == other.getAlloced() &&
             this->getStartAddr() == other.getStartAddr());
}

int FreeBlock::getSize() const {
    return size;
}

void FreeBlock::setSize(int size) {
    this->size = size;
}

bool FreeBlock::getAlloced() const {
    return Alloced;
}

void FreeBlock::setAlloced(bool alloced) {
    this->Alloced = alloced;
}

int FreeBlock::getStartAddr() const {
    return startAddr;
}

void FreeBlock::setStartAddr(int startAddr) {
    this->startAddr = startAddr;
}
